File: problems_list/剑指OfferII029.排序的循环链表/剑指OfferII029-排序的循环链表.cpp
class Solution {
public:
    Node* insert(Node* head, int insertVal) {
        //空的，没啥好说的
        if (!head) {
            Node* ret = new Node(insertVal);
            ret->next = ret;
            return ret;
        }
        //插入位置就两种，一种是中间，另一种是两侧。
        //先寻找两侧的值，在确定尾结点值得同时如果中间位置能插入的话那就插入返回就完事了
        Node* p = head->next;
        Node* p_pre = head;
        while (p != head) {
            if (
                (p_pre->val <= insertVal && insertVal <= p->val) ||  //情况1
                (p->val < p_pre->val && (insertVal <= p->val || insertVal >= p_pre->val))  //情况2
            ) {
                p_pre->next = new Node(insertVal, p);
                return head;
            }
            p_pre = p;
            p = p->next;
        }
        //两侧
        p_pre->next = new Node(insertVal, head);  //注意是p_pre才是链表的最后一个节点
        return head;
    }
};

File: problems_list/剑指OfferII029.排序的循环链表/剑指OfferII029-排序的循环链表.js
/**
 * // Definition for a Node.
 * function Node(val, next) {
 *     this.val = val;
 *     this.next = next;
 * };
 */

/**
 * @param {Node} head
 * @param {number} insertVal
 * @return {Node}
 */
var insert = function(head, insertVal) {
    // 0 节点
    if(head===null){
        let node=new Node(insertVal, null);
        node.next=node;
        return node;
    }

    // 列表中，列表首尾
    let cur=head;
    while(true){
        let nxt=cur.next;
        if(cur.val<=nxt.val && insertVal>=cur.val && insertVal<=nxt.val){
            let node=new Node(insertVal, nxt);
            cur.next=node;
            break;
        }else if(cur.val>nxt.val && (insertVal>cur.val || insertVal<nxt.val)){
            let node=new Node(insertVal, nxt);
            cur.next=node;
            break;
        }else if(cur.val===nxt.val && nxt===head){
            let node=new Node(insertVal, nxt);
            cur.next=node;
            break;
        }
        cur=cur.next;
    }
    return head;
};